src/main/ruby/truffleruby/core/struct.rb

```diff
@@ -33,7 +33,16 @@ class << self
     alias_method :subclass_new, :new
   end
 
-  def self.new(klass_name, *attrs, keyword_init: false, &block)
+  NOT_PROVIDED = BasicObject.new
```

Comment:
Thanks! Using BasicObject was indeed ugly, can you expand on how to use "undefined" ? I can't do it, a literal mention of "undefined" is just a NameError.

Anyway, I noticed that CRuby considers passing "keyword_init: nil" to Struct.new as if keyword_init was omitted (i.e. keyword_init? returns nil in this case, as if keyword_init was never passed), so I had the much simpler idea to just initialize keyword_init to nil by default : every boolean test on keyword_init is unaffected because nil is falsey, but we can still distinguish the ommited case by Primitive.nil?. Much more elegant than what I had before.

Reply:
Indeed, we don't need to distinguish cases when `keyword_init` isn't passed and when passed `nil.

Regarding `undefined` - it's a magic object, that is available by default only in Ruby Core classes (and it shouldn't be available in irb, for instance). It's widely used in the Core library:

https://github.com/oracle/truffleruby/blob/01e0ce54977a4acd79c63a52ba576f08e2121b6d/src/main/ruby/truffleruby/core/io.rb#L430-L435